skiko/src/commonMain/kotlin/org/jetbrains/skiko/context/ContextHandler.kt

```diff
@@ -5,41 +5,33 @@ import org.jetbrains.skia.Canvas
 import org.jetbrains.skia.DirectContext
 import org.jetbrains.skia.Picture
 import org.jetbrains.skia.Surface
-import org.jetbrains.skiko.GraphicsApi
-import org.jetbrains.skiko.SkiaLayer
-import org.jetbrains.skiko.hostArch
-import org.jetbrains.skiko.hostOs
-import org.jetbrains.skiko.OS
-
-internal abstract class ContextHandler(val layer: SkiaLayer) {
-    open val clearColor = if (layer.transparency || hostOs == OS.MacOS) 0 else -1
-    var context: DirectContext? = null
-    var renderTarget: BackendRenderTarget? = null
-    var surface: Surface? = null
-    var canvas: Canvas? = null
-
-    abstract fun initContext(): Boolean
-    abstract fun initCanvas()
-
-    fun clearCanvas() {
-        val color = if (layer.fullscreen && hostOs != OS.MacOS) -1 else clearColor
-        canvas?.clear(color)
-    }
-
-    open fun drawOnCanvas(picture: Picture) {
-        canvas?.drawPicture(picture)
-    }
-
-    open fun flush() {
+import org.jetbrains.skiko.*
+import org.jetbrains.skiko.RenderException
+
+internal abstract class ContextHandler(
+    protected val layer: SkiaLayer,
+    private val drawContent: (Canvas) -> Unit
+) {
+    // TODO can we simplify clearColor logic? is there a reason why SoftwareContextHandler has opposite logic?
+    protected open val clearColor = if (layer.transparency || hostOs == OS.MacOS) 0 else -1
```

Comment:
Can we group all mutable nullable fields to class, i.e. `State` with val non-null fields and init/dispose them together.

Reply:
It won't simplify the code. Grouping into another class make sense, if we not only move the state into the group, but also operations for it (to avoid !! and ?, and make operations robust and thread safe).

And this is a huge change and out of scope of the PR.

Reply:
>It won't simplify the code

BTW, I tried it, and code become unbeautiful :)

Reply:
Usually this transformation results in better code. Maybe do that myself later on.